File: bank-ocr/csharp/src/BankOcr/Digit.cs
namespace BankOcr;

public readonly record struct Digit
{
    public int? Value { get; }

    private Digit(int? value) { Value = value; }

    public static Digit Of(int value)
    {
        if (value < 0 || value > 9)
            throw new ArgumentOutOfRangeException(nameof(value), "Digit value must be 0-9.");
        return new Digit(value);
    }

    public static Digit Unknown { get; } = new(null);

    public bool IsKnown => Value.HasValue;

    public override string ToString() => IsKnown ? Value!.Value.ToString() : "?";
}

File: bank-ocr/typescript/src/Digit.ts
export class Digit {
  private constructor(public readonly value: number | null) {}

  static of(value: number): Digit {
    if (value < 0 || value > 9 || !Number.isInteger(value)) {
      throw new RangeError('Digit value must be an integer 0-9.');
    }
    return new Digit(value);
  }

  static readonly unknown: Digit = new Digit(null);

  get isKnown(): boolean { return this.value !== null; }

  equals(other: Digit): boolean { return this.value === other.value; }

  toString(): string { return this.isKnown ? String(this.value) : '?'; }
}
